node/impl/full.go

```diff
@@ -118,6 +125,101 @@ func (n *FullNodeAPI) NodeStatus(ctx context.Context, inclChainStatus bool) (sta
 	return status, nil
 }
 
+// This method is exclusively used for Non Interactive Authentication based on SP Worker Address
+func (n *FullNodeAPI) FilIdSp(ctx context.Context, spID address.Address, optionalArg []byte) (string, error) {
+
+	head, err := n.ChainHead(ctx)
+	if err != nil {
+		return "", xerrors.Errorf("failed to get chain head: %w", err)
+	}
+
+	finHeight := head.Height() - miner.ChainFinality
+	finTipSet, err := n.ChainGetTipSetByHeight(ctx, finHeight, head.Key())
+	if err != nil {
+		return "", xerrors.Errorf("failed to get finalized tipset at %d: %w", finHeight, err)
+	}
+
+	// Get miner info for last finality
+	minfo, err := n.StateMinerInfo(ctx, spID, finTipSet.Key())
+	if err != nil {
+		return "", xerrors.Errorf("failed to get info for SP %s at %d: %w", spID.String(), finTipSet.Height(), err)
+	}
+
+	return signNoninteractiveChallege(
+		ctx,
+		n,
+		"FIL-SPID-V0",
+		spID, minfo.Worker,
+		head.Height(),
+		optionalArg,
+	)
+}
+
+// This method is exclusively used for Non Interactive Authentication based on provided Wallet Address
+func (n *FullNodeAPI) FilIdAddr(ctx context.Context, addr address.Address, optionalArg []byte) (string, error) {
+
+	head, err := n.ChainHead(ctx)
+	if err != nil {
+		return "", xerrors.Errorf("failed to get chain head: %w", err)
+	}
+
+	if addr.Protocol() == address.ID {
+		resolvedAddr, err := n.StateAccountKey(ctx, addr, head.Key())
+		if err != nil {
+			return "", xerrors.Errorf("could not resolve ID address %s to public address: %w", addr.String(), err)
+		}
+		addr = resolvedAddr
+	}
+
+	return signNoninteractiveChallege(
+		ctx,
+		n,
+		"FIL-ADDRID-V0",
+		addr, addr,
+		head.Height(),
+		optionalArg,
+	)
+}
+
+func signNoninteractiveChallege(ctx context.Context, n *FullNodeAPI, challengeType string, target, signer address.Address, drandHeight abi.ChainEpoch, optionalArg []byte) (string, error) {
+
+	if len(optionalArg) > 2048 {
+		return "", xerrors.Errorf("optional argument longer than the maximum of 2048 bytes")
+	}
+
+	beacon, err := n.StateGetBeaconEntry(ctx, drandHeight)
+	if err != nil {
+		return "", xerrors.Errorf("failed to get DRAND beacon: %w", err)
+	}
+
+	sig, err := n.WalletSign(
+		ctx,
+		signer,
+		append(
+			append(
+				append(
+					// Prefix is to avoid ever producing a viable CBOR message
+					// FIXME: this should be replaced by SPIDV1 combined with proper domain-separation when it arrives
+					make([]byte, 0, 3+len(beacon.Data)+len(optionalArg)),
+					[]byte("   ")...,
+				),
+				beacon.Data...,
+			),
+			optionalArg...,
+		),
```

Comment:
I'd break this into separate lines, now it's not easy to tell what data this will actually output

Reply:
Also I'd expect target addr to be part of signed data

Reply:
> Also I'd expect target addr to be part of signed data

What additional guarantee would this achieve?